### ISSUE TITLE
Reduce memory consumption

### ISSUE THREAD

**Issue by alebar** [micrometer-metrics/micrometer#516]
We observed increased memory usage in our microservices, after applying micrometer. Heap dumps showed that a lot of memory is consumed by histograms stored in our timers. Following is an extract from such dump:

```
io.micrometer.core.instrument.step.StepTimer@0xf97e62a0
- histogram: TimeWindowLatencyHistogram@0xf97e6398
  - accumulatedHistogram: Histogram@0xf982e7a0
    - counts: [long]@0xf982e9f0 (10256 bytes) 
  - ringBuffer: [LatencyStats]
    0: LatencyStats@0xf9821a50
      - activePauseCorrectionsHistogram: Histogram@0xf98299f0
        - counts: [long]@0xf9829c40 (10256 bytes) 
      - activeRecordingHistogram: AtomicHistogram@0xf9826f70
        - counts: AtomicHistogram@0xf98271d0
          - array: [long]@0xf98271e0 (10256 bytes)
      - inactivePauseCorrectionsHistogram: Histogram@0xf9824510
        - counts: [long]@0xf9824760 (10256 bytes)
      - inactiveRawDataHistogram: AtomicHistogram@0xf9821a90
        - counts: AtomicHistogram@0xf9821cf0
          - array: [long]@0xf9821d00 (10256 bytes)
      - intervalEstimator: TimeCappedMovingAverageIntervalEstimator@0xf982c4c8
        - intervalEndTimes: [long]@0xf982c518 (8208 bytes) 
```
And there is 4 more LatencyStats objects in that ringBuffer array.

Summing only long arrays:
10256 + (10256 * 4) * 5 = 215376 bytes
Total for all 80 timers that we have:
215376 * 80 = 16.4 mb

We don't use histograms and percentiles, so we've overcame the problem with custom implementation of the Timer interface. But it would be great to have a better solution straight in the Micrometer.

**Comment by jkschneider** [comment]
Reducing the digits of precision to 0 on the underlying hdr histogram (which it should be anyway, since we are always dimensioning that in nanos) reduces the footprint of each histogram from 10,752 bytes to 648. i think we can also reduce the default ring buffer length from 5 to 3. so 10752*5 becomes 648*3.

Also we'll plan to drop histograms altogether when percentiles aren’t being used.

**Comment by jkschneider** [comment]
After decreasing buffer length to 3, adding a new configuration option for digits of precision used in computing percentile approximations, and adding new histogram implementations that are more memory efficient when you don't need client-side percentiles, we have this result:

* No histogram, no percentiles: 0.56kb/timer
* Histogram, no percentiles: 0.83kb/timer
* Percentiles, 1 digit precision: 3.1kb/timer    (default precision in 1.0.3)
* Percentiles, 2 digits precision: 11.6kb/timer  (default precision in 1.0.2)
* Percentiles, 3 digits precision: 58.7kb/timer

**Comment by jkschneider** [comment]
Fixed with 535b2aa4.

Little bit more polish to do, but this represents much of the work I think.

**Comment by alebar** [comment]
Great! That looks awesome. Thank you.

Just one question: how the histograms can be turned on or off? Using the `MeterFilter`?

**Comment by jkschneider** [comment]
@alebar The code is now conditionalized to use one of three `Histogram` implementations depending on what features you select. A `NoopHistogram` which is essentially free, `TimeWindowFixedBoundaryHistogram` for percentile histograms/SLAs which is more expensive, and the `TimeWindowPercentileHistogram` for client-side percentiles which is the most expensive.

We never configure SLAs, percentile histograms, or client-side percentiles for you in Spring's autoconfiguration. It is up to you to attach them to timers through [per-meter filter properties](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#_per_meter_properties), a programmatically defined `MeterFilter`, or through the `Timer.Builder` class for custom timers.

**Comment by alebar** [comment]
Ok. Thanks

**Comment by pooyas** [comment]
Is there an alternative to access removed public methods? e.g percentile in Timer?

**Comment by jkschneider** [comment]
@pooyas Sorry about the breaking change -- I have some heartburn about this in a non-major for sure.

You can access preconfigured client-side percentiles through `timer.takeSnapshot().percentileValues()`

In general, this method should never have been on `Timer` since some histogram implementations do not support percentiles at all.

**Comment by pooyas** [comment]
@jkschneider thanks for proposing the alternative. I'm using it to get percentile value of a time distribution for a defined time window.

**Comment by jkschneider** [comment]
I think we should go ahead and put it back with a NaN return when the Timer isn't preconfigured with percentiles.

**Comment by pooyas** [comment]
@jkschneider that would be great, it makes it backward compatible at least on interface-level

**Comment by jkschneider** [comment]
@pooyas Fixed in fa78620f. Thanks for the timely feedback.

**Comment by pooyas** [comment]
@jkschneider thanks!
